### PR TITLE
[ENH] Add capacities to prepare Topup and Eddy

### PR DESCRIPTION
**Pull request by boiteaclou**
Modifications to prepare_topup :
- Handles b0 images composed of multiple forward and reverse phase acquired volumes
- Added parameter to change the name of the acquisition parameters file outputed
- IMPORTANT : the script does not accept a DWI as input for the forward acquired image, only a b0 volume. Users must handle the extraction of the b0s using scil_extract_b0

Modifications to prepare_eddy :
- Added logic to handle the different sizes of b0 images that Topup ran on
- Added parameters to input the number of reverse phase volumes present in the input DWI image
- Added parameter that takes a acquisition parameters file as an input, required to know the number of b0 volumes to take into account when creating the index file
-  Added parameter to get least-square resampling outputs when executing Eddy, as well as checks on the b-values and b-vectors to ensure such resampling is possible

### CODE REVIEW

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_prepare_eddy_command.py [hunk 1]
```diff
@@ -144,6 +183,19 @@ def main():
     if args.fix_seed:
         additional_args += "--initrand "
 
+    if args.lsr_resampling:
+        if len(bvals) - n_rev == n_rev:
+            forward_bb = bvals[:n_rev, None] * bvecs[:n_rev, :]
+            reverse_bb = bvals[n_rev:, None] * bvecs[n_rev:, :]
+            if np.allclose(forward_bb, reverse_bb):
+                additional_args += "--resamp lsr "
```
Comment (arnaudbore) [inline scripts/scil_prepare_eddy_command.py:191]:
```suggestion
                additional_args += "--resamp=lsr --fep=true "
```

**Review by arnaudbore — COMMENTED** [review]
Really good job ! A couple of comments now I need to find if there is a black hole somewhere. ;)

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

File: scilpy/tractanalysis/todi.py [hunk 2]
```diff
@@ -200,7 +200,8 @@ def smooth_todi_spatial(self, sigma=0.5):
                 new_todi = deepcopy(tmp_todi)
```
Comment (GuillaumeTh) [inline scilpy/tractanalysis/todi.py:200]: COuld you remove this file.

File: scripts/scil_generate_priors_from_bundle.py [hunk 3]
```diff
@@ -12,13 +12,14 @@
 import os
```
Comment (GuillaumeTh) [inline scripts/scil_generate_priors_from_bundle.py:12]: COuld you remove this file.

File: scripts/scil_score_tractogram.py [hunk 4]
```diff
@@ -150,7 +150,13 @@ def main():
```
Comment (GuillaumeTh) [inline scripts/scil_score_tractogram.py:150]: COuld you remove this file.

File: Jenkinsfile [hunk 5]
```diff
@@ -4,21 +4,11 @@ pipeline {
     stages {
```
Comment (GuillaumeTh) [inline Jenkinsfile:4]: COuld you remove this file.

**Review by GuillaumeTh — APPROVED** [review]

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Please fix conflicts + put a warning when `--n_reverse` different from 0 and `not --lsr_resampling`

File: scripts/scil_prepare_topup_command.py [hunk 6]
```diff
@@ -24,30 +24,19 @@ def _build_arg_parser():
...
-                   help='b-values file in FSL format')
-
-    p.add_argument('in_bvecs',
-                   help='b-vectors file in FSL format')
+    p.add_argument('in_forward_b0',
+                   help='input b0 Nifti image with forward phase encoding')
 
     p.add_argument('in_reverse_b0',
-                   help='b0 image with reversed phase encoding.')
+                   help='input b0 Nifti image with reversed phase encoding.')
```
Comment (arnaudbore) [inline scripts/scil_prepare_topup_command.py:31]:
```suggestion
                   help='Input b0 Nifti image with reversed phase encoding.')
```

File: scripts/scil_prepare_topup_command.py [hunk 7]
```diff
@@ -24,30 +24,19 @@ def _build_arg_parser():
...
 
     p.add_argument('--config', default='b02b0.cnf',
                    help='topup config file [%(default)s].')
```
Comment (arnaudbore) [inline scripts/scil_prepare_topup_command.py:34]:
```suggestion
                   help='Topup config file [%(default)s].')
```

File: scripts/scil_prepare_topup_command.py [hunk 8]
```diff
@@ -24,30 +24,19 @@ def _build_arg_parser():
...
-
     p.add_argument('--encoding_direction', default='y',
                    choices=['x', 'y', 'z'],
-                   help='acquisition direction, default is AP-PA '
-                        '[%(default)s].')
+                   help='acquisition direction of the forward b0 '
+                        'image, default is AP [%(default)s].')
 
     p.add_argument('--readout', type=float, default=0.062,
                    help='total readout time from the DICOM metadata '
```
Comment (arnaudbore) [inline scripts/scil_prepare_topup_command.py:42]:
```suggestion
                   help='Total readout time from the DICOM metadata '
```

File: scripts/scil_prepare_topup_command.py [hunk 9]
```diff
@@ -62,6 +51,10 @@ def _build_arg_parser():
     p.add_argument('--out_prefix', default='topup_results',
                    help='prefix of the topup results [%(default)s].')
 
+    p.add_argument('--out_params', default='acqparams.txt',
+                   help='filename for the acquisition '
```
Comment (arnaudbore) [inline scripts/scil_prepare_topup_command.py:55]:
```suggestion
                   help='Filename for the acquisition '
```

File: scripts/scil_prepare_topup_command.py [hunk 10]
```diff
@@ -24,30 +24,19 @@ def _build_arg_parser():
...
-    p.add_argument('in_dwi',
-                   help='input DWI Nifti image')
-
-    p.add_argument('in_bvals',
-                   help='b-values file in FSL format')
-
-    p.add_argument('in_bvecs',
-                   help='b-vectors file in FSL format')
+    p.add_argument('in_forward_b0',
+                   help='input b0 Nifti image with forward phase encoding')
```
Comment (arnaudbore) [inline scripts/scil_prepare_topup_command.py:28]:
```suggestion
                   help='Input b0 Nifti image with forward phase encoding.')
```

**Review by arnaudbore — COMMENTED** [review]
I was not able to put all the dots missing or capital letter at the beginning of helps but you understand the issue. For prepare_eddy for in_dwi please add in the help saying something like this: if multiple acquisitions please merge in the same order as prepare_topup using scil_concatenate_dwi.py. Could also be added in the description for more detailed explanations.

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Review by arnaudbore — APPROVED** [review]
Go !! Great work @boiteaclou

**Pull request merged**